### PR TITLE
feat(telemetry): server-side usage statistics

### PR DESCRIPTION
**Pull request by kevinhu**
## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/ScheduledAnalyticsFactory.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+            // set user-level properties
+            JSONObject report = new JSONObject();
+            report.put("dau", dailyActiveUsers);
+            report.put("wau", weeklyActiveUsers);
+            report.put("mau", monthlyActiveUsers);
+
+            Telemetry.ping("service-daily", report);
+
+        } catch (Exception e) {
```
Comment (dexter-mh-lee) [inline metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/ScheduledAnalyticsFactory.java:73]: What kind of exceptions are we catching here?

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/ScheduledAnalyticsFactory.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            log.error("Error computing daily report:\n" + ExceptionUtils.getStackTrace(e));
```
Comment (dexter-mh-lee) [inline metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/ScheduledAnalyticsFactory.java:74]: for log.error, you can just do
log.error("Error computing daily report", e)

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/ScheduledAnalyticsFactory.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
+
+import java.util.Optional;
+
+@Slf4j
+@Configuration
+@EnableScheduling
+@RequiredArgsConstructor
```
Comment (dexter-mh-lee) [inline metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/ScheduledAnalyticsFactory.java:25]: Dont think you need this

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/Telemetry.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+import java.util.UUID;
+
+import lombok.extern.slf4j.Slf4j;
+
+import static java.lang.Boolean.parseBoolean;
+
+@Slf4j
+final class Telemetry {
+    static final String MIXPANEL_TOKEN = "5ee83d940754d63cacbf7d34daa6f44a";
+    static final String DATAHUB_FOLDER_PATH = System.getProperty("user.home") + "/.datahub";
```
Comment (dexter-mh-lee) [inline metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/Telemetry.java:23]: I think these paths are wrong? Are these not finalized yet? 
Ideally we should take this path as a parameter

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/Telemetry.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+                // set user-level properties
+                JSONObject props = new JSONObject();
+                props.put("java_version", System.getProperty("java.version"));
+                props.put("os", System.getProperty("os.name"));
+                JSONObject update = mixpanelBuilder.set(clientId, props);
+                mixpanel.sendMessage(update);
+            }
+
+        } catch (Exception e) {
```
Comment (dexter-mh-lee) [inline metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/Telemetry.java:58]: catching whole chunk of code is not recommended. What are we catching here? If it's just whether the file exists on the path, we should try catch just that

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/ScheduledAnalyticsFactory.java (hunk 1 above)
Comment (kevinhu) [reply metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/ScheduledAnalyticsFactory.java]: Fixed!

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/Telemetry.java (hunk 4 above)
Comment (kevinhu) [reply metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/Telemetry.java]: Fixed!

**Comment by anshbansal** [comment]
For server side we recently added more granular env variables https://github.com/datahub-project/datahub/blob/master/metadata-service/factories/src/main/resources/application.yml#L32 which our server needs to respect.

This is how we are getting them to send via the gms config https://github.com/datahub-project/datahub/blob/master/metadata-service/servlet/src/main/java/com/datahub/gms/servlet/Config.java#L82 which you can use as an example.

You can discuss with @shirshanka whether `DATAHUB_TELEMETRY_ENABLED` should be used as an overall switch on GMS for all telemetry or keep `DATAHUB_TELEMETRY_ENABLED` as a CLI only env variable. If we decide to use `DATAHUB_TELEMETRY_ENABLED` as a server side variable it needs to be enabled by default and added in `application.yml` as well be sent through the `/gms/config` endpoint. 

We need to respect server side controls in the CLI.

**Comment by kevinhu** [comment]
I see—I will sync and see what we want to do with `DATAHUB_TELEMETRY_ENABLED`. Thanks for the tips!

File: metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/ScheduledAnalyticsFactory.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    @Autowired
+    @Qualifier("elasticSearchRestHighLevelClient")
+    private RestHighLevelClient elasticClient;
+
+    @Autowired
+    @Qualifier(IndexConventionFactory.INDEX_CONVENTION_BEAN)
+    private IndexConvention indexConvention;
+
+    // statistics to send daily
+    @Scheduled(fixedDelay = 24 * 60 * 60 * 1000)
```
Comment (RyanHolstien) [inline metadata-service/factories/src/main/java/com/linkedin/gms/factory/telemetry/ScheduledAnalyticsFactory.java:35]: This should probably be configurable

**Review by RyanHolstien — APPROVED** [review]
Overall this LGTM, but I think the schedule should be a configuration option. Also needs some checkstyle fixes & merge conflict resolutions

**Comment by shirshanka** [comment]
closing as this is already done

**Pull request closed without merging**